planner/core/rule_derive_topn_from_window.go

```diff
@@ -71,12 +96,21 @@ func windowIsTopN(p *LogicalSelection) (bool, uint64) {
 	}
 
 	grandChild := child.Children()[0]
-	_, isDataSource := grandChild.(*DataSource)
+	dataSource, isDataSource := grandChild.(*DataSource)
 	if !isDataSource {
 		return false, 0
 	}
-	if len(child.WindowFuncDescs) == 1 && child.WindowFuncDescs[0].Name == "row_number" && len(child.PartitionBy) == 0 &&
-		child.Frame.Type == ast.Rows && child.Frame.Start.Type == ast.CurrentRow && child.Frame.End.Type == ast.CurrentRow {
+
+	// Give up if TiFlash is one possible access path. Pushing down window aggregation is good enough in this case.
+	for _, path := range dataSource.possibleAccessPaths {
+		if path.StoreType == kv.TiFlash {
+			return false, 0
```

Comment:
Do you mean we will not compare the cost of mpp window aggregation to the topn pushdown in this PR?

Reply:
Yes. Our goal in this PR is to allow KiTv TopN push down for now with a flag. We need a more fundamental way of choosing between mpp and TiKV later on.